openedx/api.py

```diff
@@ -127,6 +127,81 @@ def _extract_username_suggestions(data, suggestions_extracted):
     return [], suggestions_extracted
 
 
+def _generate_unique_username(base_username, max_length=OPENEDX_USERNAME_MAX_LEN):
+    """
+    Generate a unique username by appending random numbers to the base username.
+
+    Args:
+        base_username (str): The base username to use
+        max_length (int): Maximum length for the generated username
+
+    Returns:
+        str or None: A unique username, or None if no unique username could be generated
+    """
+    ranges = (
+        (0, 9),
+        (10, 99),
+        (100, 999),
+        (1000, 9999),
+        (10000, 99999),
+    )
+
+    for intrange in ranges:
+        random_int = random.randint(intrange[0], intrange[1])  # noqa: S311
+        username_to_try = f"{base_username}_{random_int}"
+
+        if len(username_to_try) > max_length:
+            amount_to_truncate = len(username_to_try) - max_length
+            username_to_try = f"{base_username[:amount_to_truncate]}-{random_int}"
+
+        if not OpenEdxUser.objects.filter(edx_username=username_to_try).exists():
+            return username_to_try
+
+    return None
+
+
+def _handle_username_collision(  # noqa: PLR0913
+    resp, data, open_edx_user, user, suggested_usernames, suggestions_extracted
+):
+    """
+    Handle username collision by trying OpenEdX suggestions or falling back to local generation.
+
+    Args:
+        resp: HTTP response from OpenEdX
+        data: Parsed JSON response data
+        open_edx_user: OpenEdxUser instance
+        user: User instance
+        suggested_usernames: List of suggested usernames from previous attempts
+        suggestions_extracted: Boolean indicating if suggestions were already extracted
+
+    Returns:
+        tuple: (new_username, should_continue, should_reset_attempts)
+    """
+    if not _is_duplicate_username_error(resp, data):
+        return None, False, False
+
+    suggestions, suggestions_extracted = _extract_username_suggestions(
+        data, suggestions_extracted
+    )
+    if suggestions:
+        suggested_usernames = suggestions
+
+    if not suggested_usernames:
+        log.info(
+            "OpenEdX returned empty username suggestions, falling back to local generation"
+        )
+        base_username = open_edx_user.desired_edx_username or user.username
```

Comment:
`User.username` will be the email address, but I don't think we want to use the entire thing because I think openedx rejects usernames with `@` in them. We could use the local part though, but I would also probably access this through `email` just so it's clear to readers of this code why the split on `@` is happening.